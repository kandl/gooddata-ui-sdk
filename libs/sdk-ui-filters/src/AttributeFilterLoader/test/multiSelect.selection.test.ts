// (C) 2019-2022 GoodData Corporation
import { newTestAttributeFilterHandler } from "./fixtures";
import { waitForAsync } from "./testUtils";

describe("MultiSelectAttributeFilterHandler", () => {
    it("should initialize selection and trigger callbacks for positive attribute filter", async () => {
        const onSelectionChanged = jest.fn();
        const onSelectionCommitted = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.onSelectionChanged(onSelectionChanged);
        attributeFilterHandler.onSelectionCommitted(onSelectionCommitted);

        attributeFilterHandler.init();

        await waitForAsync();

        expect(onSelectionChanged).toHaveBeenCalledTimes(1);
        expect(onSelectionCommitted).toHaveBeenCalledTimes(1);
        expect(onSelectionChanged.mock.calls[0]).toMatchSnapshot("onSelectionChanged parameters");
        expect(onSelectionCommitted.mock.calls[0]).toMatchSnapshot("onSelectionCommitted parameters");
        expect(attributeFilterHandler.getCommittedSelection()).toMatchSnapshot("committed selection");
        expect(attributeFilterHandler.getWorkingSelection()).toMatchSnapshot("working selection");
    });

    it("should initialize selection and trigger callbacks for negative attribute filter", async () => {
        const onSelectionChanged = jest.fn();
        const onSelectionCommitted = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("negative");

        attributeFilterHandler.onSelectionChanged(onSelectionChanged);
        attributeFilterHandler.onSelectionCommitted(onSelectionCommitted);

        attributeFilterHandler.init();

        await waitForAsync();

        expect(onSelectionChanged).toHaveBeenCalledTimes(1);
        expect(onSelectionCommitted).toHaveBeenCalledTimes(1);
        expect(onSelectionChanged.mock.calls[0]).toMatchSnapshot("onSelectionChanged parameters");
        expect(onSelectionCommitted.mock.calls[0]).toMatchSnapshot("onSelectionCommitted parameters");
        expect(attributeFilterHandler.getCommittedSelection()).toMatchSnapshot("committed selection");
        expect(attributeFilterHandler.getWorkingSelection()).toMatchSnapshot("working selection");
    });

    it("should change selection and trigger callbacks", async () => {
        const onSelectionChanged = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.init();

        await waitForAsync();

        attributeFilterHandler.onSelectionChanged(onSelectionChanged);
        attributeFilterHandler.changeSelection({
            isInverted: true,
            items: ["/gdc/md/referenceworkspace/obj/1054/elements?id=166497"],
        });

        expect(onSelectionChanged).toHaveBeenCalledTimes(1);
        expect(onSelectionChanged.mock.calls[0]).toMatchSnapshot("onSelectionChanged parameters");
        expect(attributeFilterHandler.getWorkingSelection()).toMatchSnapshot("working selection");
    });

    it("should revert selection and trigger callbacks", async () => {
        const onSelectionChanged = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.init();

        await waitForAsync();

        expect(attributeFilterHandler.getWorkingSelection()).toMatchSnapshot(
            "working selection before revert",
        );

        attributeFilterHandler.changeSelection({
            isInverted: true,
            items: ["/gdc/md/referenceworkspace/obj/1054/elements?id=166497"],
        });

        expect(attributeFilterHandler.getWorkingSelection()).toMatchSnapshot(
            "working selection after change",
        );

        attributeFilterHandler.onSelectionChanged(onSelectionChanged);
        attributeFilterHandler.revertSelection();

        expect(onSelectionChanged).toHaveBeenCalledTimes(1);
        expect(onSelectionChanged.mock.calls[0]).toMatchSnapshot("onSelectionChanged callback parameters");
        expect(attributeFilterHandler.getWorkingSelection()).toMatchSnapshot(
            "working selection after revert",
        );
    });

    it("should invert selection and trigger callbacks", async () => {
        const onSelectionChanged = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.init();

        await waitForAsync();

        expect(attributeFilterHandler.getWorkingSelection()).toMatchSnapshot(
            "working selection before invert",
        );

        attributeFilterHandler.onSelectionChanged(onSelectionChanged);
        attributeFilterHandler.invertSelection();

        expect(onSelectionChanged).toHaveBeenCalledTimes(1);
        expect(onSelectionChanged.mock.calls[0]).toMatchSnapshot("onSelectionChanged parameters");
        expect(attributeFilterHandler.getWorkingSelection()).toMatchSnapshot(
            "working selection after invert",
        );
    });

    it("should commit selection and trigger callbacks", async () => {
        const onSelectionCommitted = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.init();

        await waitForAsync();

        attributeFilterHandler.onSelectionCommitted(onSelectionCommitted);
        attributeFilterHandler.changeSelection({
            isInverted: true,
            items: ["/gdc/md/referenceworkspace/obj/1054/elements?id=166497"],
        });
        attributeFilterHandler.commitSelection();

        expect(onSelectionCommitted).toHaveBeenCalledTimes(1);
        expect(onSelectionCommitted.mock.calls[0]).toMatchSnapshot("onSelectionCommitted parameters");
        expect(attributeFilterHandler.getCommittedSelection()).toMatchSnapshot("committed selection");
    });

    it("should clear selection and trigger callbacks", async () => {
        const onSelectionChanged = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.init();

        await waitForAsync();

        attributeFilterHandler.onSelectionChanged(onSelectionChanged);
        attributeFilterHandler.clearSelection();

        expect(onSelectionChanged).toHaveBeenCalledTimes(1);
        expect(onSelectionChanged.mock.calls[0]).toMatchSnapshot("onSelectionChanged parameters");
        expect(attributeFilterHandler.getWorkingSelection()).toMatchSnapshot("working selection");
    });
});
