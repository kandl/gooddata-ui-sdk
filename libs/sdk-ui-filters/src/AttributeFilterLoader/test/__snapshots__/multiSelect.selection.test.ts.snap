// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelectAttributeFilterHandler should change selection and trigger callbacks: onSelectionChanged parameters 1`] = `
Array [
  Object {
    "selection": Object {
      "isInverted": true,
      "items": Array [
        "/gdc/md/referenceworkspace/obj/1054/elements?id=166497",
      ],
    },
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should change selection and trigger callbacks: working selection 1`] = `
Object {
  "isInverted": true,
  "items": Array [
    "/gdc/md/referenceworkspace/obj/1054/elements?id=166497",
  ],
}
`;

exports[`MultiSelectAttributeFilterHandler should clear selection and trigger callbacks: onSelectionChanged parameters 1`] = `
Array [
  Object {
    "selection": Object {
      "isInverted": false,
      "items": Array [],
    },
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should clear selection and trigger callbacks: working selection 1`] = `
Object {
  "isInverted": false,
  "items": Array [],
}
`;

exports[`MultiSelectAttributeFilterHandler should commit selection and trigger callbacks: committed selection 1`] = `
Object {
  "isInverted": true,
  "items": Array [
    "/gdc/md/referenceworkspace/obj/1054/elements?id=166497",
  ],
}
`;

exports[`MultiSelectAttributeFilterHandler should commit selection and trigger callbacks: onSelectionCommitted parameters 1`] = `
Array [
  Object {
    "selection": Object {
      "isInverted": true,
      "items": Array [
        "/gdc/md/referenceworkspace/obj/1054/elements?id=166497",
      ],
    },
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should initialize selection and trigger callbacks for negative attribute filter: committed selection 1`] = `
Object {
  "isInverted": false,
  "items": Array [
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
  ],
}
`;

exports[`MultiSelectAttributeFilterHandler should initialize selection and trigger callbacks for negative attribute filter: onSelectionChanged parameters 1`] = `
Array [
  Object {
    "selection": Object {
      "isInverted": false,
      "items": Array [
        "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
        "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
      ],
    },
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should initialize selection and trigger callbacks for negative attribute filter: onSelectionCommitted parameters 1`] = `
Array [
  Object {
    "selection": Object {
      "isInverted": false,
      "items": Array [
        "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
        "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
      ],
    },
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should initialize selection and trigger callbacks for negative attribute filter: working selection 1`] = `
Object {
  "isInverted": false,
  "items": Array [
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
  ],
}
`;

exports[`MultiSelectAttributeFilterHandler should initialize selection and trigger callbacks for positive attribute filter: committed selection 1`] = `
Object {
  "isInverted": false,
  "items": Array [
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
  ],
}
`;

exports[`MultiSelectAttributeFilterHandler should initialize selection and trigger callbacks for positive attribute filter: onSelectionChanged parameters 1`] = `
Array [
  Object {
    "selection": Object {
      "isInverted": false,
      "items": Array [
        "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
        "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
      ],
    },
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should initialize selection and trigger callbacks for positive attribute filter: onSelectionCommitted parameters 1`] = `
Array [
  Object {
    "selection": Object {
      "isInverted": false,
      "items": Array [
        "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
        "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
      ],
    },
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should initialize selection and trigger callbacks for positive attribute filter: working selection 1`] = `
Object {
  "isInverted": false,
  "items": Array [
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
  ],
}
`;

exports[`MultiSelectAttributeFilterHandler should invert selection and trigger callbacks: onSelectionChanged parameters 1`] = `
Array [
  Object {
    "selection": Object {
      "isInverted": true,
      "items": Array [
        "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
        "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
      ],
    },
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should invert selection and trigger callbacks: working selection after invert 1`] = `
Object {
  "isInverted": true,
  "items": Array [
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
  ],
}
`;

exports[`MultiSelectAttributeFilterHandler should invert selection and trigger callbacks: working selection before invert 1`] = `
Object {
  "isInverted": false,
  "items": Array [
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
  ],
}
`;

exports[`MultiSelectAttributeFilterHandler should revert selection and trigger callbacks: onSelectionChanged callback parameters 1`] = `
Array [
  Object {
    "selection": Object {
      "isInverted": false,
      "items": Array [
        "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
        "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
      ],
    },
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should revert selection and trigger callbacks: working selection after change 1`] = `
Object {
  "isInverted": true,
  "items": Array [
    "/gdc/md/referenceworkspace/obj/1054/elements?id=166497",
  ],
}
`;

exports[`MultiSelectAttributeFilterHandler should revert selection and trigger callbacks: working selection after revert 1`] = `
Object {
  "isInverted": false,
  "items": Array [
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
  ],
}
`;

exports[`MultiSelectAttributeFilterHandler should revert selection and trigger callbacks: working selection before revert 1`] = `
Object {
  "isInverted": false,
  "items": Array [
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165678",
    "/gdc/md/referenceworkspace/obj/1054/elements?id=165847",
  ],
}
`;
